File: lab/segmentintersection.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <iomanip>

/* 
Does not work.
*/

using namespace std;
using ll = long long;

class Vec2D;

// Alias types to make code clearer
using Point   = Vec2D;
using Polygon = vector<Point>;
using Segment = pair<Point, Point>;

const double EPS = 10e-9;
const double PI  = acos(-1.0);
const double INF = 1e18;

class Vec2D
{
public:
/**
 * X coord
 */
double x;
/**
 * Y coord
 */
double y;
/**
 * New vector
 * @param x X coord
 * @param y Y coord
 */
Vec2D(double x = 0, double y = 0) : x(x), y(y) {}
/**
 * Minus operator
 * @param other Vector
 */
Vec2D operator-(const Vec2D & other) const { return Vec2D(x - other.x, y - other.y); }
/**
 * Plus operator
 * @param other Vector
 */
Vec2D operator+(const Vec2D & other) const { return Vec2D(x + other.x, y + other.y); }
/**
 * Scalar multiplication operator
 * @param scalar Scalar value
 */
Vec2D operator*(double scalar) const { return Vec2D(x * scalar, y * scalar); }
/**
 * Scalar division operator
 * @param scalar Scalar value
 */
Vec2D operator/(double scalar) const {
    if (scalar == 0) {
        throw invalid_argument("Division by zero");
    }
    return Vec2D(x / scalar, y / scalar);
}
/**
 * Equality operator
 * @param other Vector
 */
bool operator==(const Vec2D & other) const {
    return abs(x - other.x) < EPS && abs(y - other.y) < EPS;
}

/**
 * Cross product
 * @param other Vector
 */
double cross(const Vec2D& other) const {
    return x * other.y - y * other.x;
}

/**
 * Dot product
 * @param other Vector
 */
double dot(const Vec2D& other) const {
    return x * other.x + y * other.y;
}

/**
 * Distance to another vector
 * @param other Vector
 */
double dist(const Vec2D& other) const {
    return sqrt((x - other.x) * (x - other.x) + (y - other.y) * (y - other.y));
}
};

/**
 * Check if point p lies on the line segment s.
 * @param p Point
 * @param s Segment
 * @return true/false
 */
bool is_on_segment(Point& p, Segment& s) {
    return (p.x >= min(s.first.x, s.second.x) - EPS &&
            p.x <= max(s.first.x, s.second.x) + EPS &&
            p.y >= min(s.first.y, s.second.y) - EPS &&
            p.y <= max(s.first.y, s.second.y) + EPS);
}

/**
 * Find intersection between two line segments
 * @param s1 First segment
 * @param s2 Second segment
 * @return Vector of intersection points (0, 1, or 2 points)
 */
vector<Point> intersect(Segment s1, Segment s2)
{
    // line 1
    Point p = s1.first;
    Vec2D r = s1.second - s1.first;

    // line 2
    Point q = s2.first;
    Vec2D s = s2.second - s2.first;

    double rXs = r.cross(s);
    Vec2D pq = q - p;

    // lines collinear if their direction vectors are parallel (rXs)
    // and vector between their starting points (pq) is parallel to any line's direction
    if (abs(rXs) < EPS && abs(pq.cross(r)) < EPS) {
        // candidate points
        vector<Point> pts;

        if (is_on_segment(s1.first,  s2)) pts.push_back(s1.first);
        if (is_on_segment(s1.second, s2)) pts.push_back(s1.second);
        if (is_on_segment(s2.first,  s1)) pts.push_back(s2.first);
        if (is_on_segment(s2.second, s1)) pts.push_back(s2.second);

        // No overlap
        if (pts.empty()) return {};

        std::sort(pts.begin(), pts.end(), [](const Point& pt1, const Point& pt2)
        {
            // Prefer X
            if (abs(pt1.x - pt2.x) > EPS) {
                return pt1.x < pt2.x;
            }

            // X are equal
            if (abs(pt1.y - pt2.y) > EPS) {
                 return pt1.y < pt2.y;
            }

            return false;
        });

        // Remove duplicate points
        pts.erase(std::unique(pts.begin(), pts.end(), [](Point pt1, Point pt2){
            return abs(pt1.x - pt2.x) < EPS && abs(pt1.y - pt2.y) < EPS;
        }), pts.end());

        // Should be impossible
        if (pts.empty()) return {};

        // Point overlap
        if (pts.size() == 1) return {pts[0]};
        
        // Segment overlap
        return {pts.front(), pts.back()};
    }

    // Lines are parallel but NOT collinear
    if (abs(rXs) < EPS) { 
        return {};
    }

    // Single point intersect
    // Solve for t and u: p + t * r = q + u * s
    // t => s1, u => s2
    double t = pq.cross(s) / rXs;
    double u = pq.cross(r) / rXs; 

    // Check intersect lies within both segments
    if (t >= 0 - EPS &&
        t <= 1 + EPS &&
        u >= 0 - EPS &&
        u <= 1 + EPS) 
    {
        Point intersection_point = p + r * t;
        return {intersection_point};
    }

    // Lines intersect, segments do not
    return {};
}

int main()
{
    cin.tie(nullptr);
    cin.sync_with_stdio(false);

    int n_cases;
    cin >> n_cases;
    for (int i = 0; i < n_cases; ++i) {
        double x1, y1, x2, y2, x3, y3, x4, y4;
        cin >> x1 >> y1 >> x2 >> y2 >> x3 >> y3 >> x4 >> y4;

        Segment s1 = {Point(x1, y1), Point(x2, y2)};
        Segment s2 = {Point(x3, y3), Point(x4, y4)};

        vector<Point> result = intersect(s1, s2);

        if (result.empty()) {
            cout << "none" << '\n';
            continue;
        }

        // Avoid -0.0
        auto abs_if_zero = [](double val) {
            if (abs(val) < 5*1e-3) return 0.0;
            return val;
        };
        
        cout << fixed << setprecision(2);
        if (result.size() == 1) {
            cout << abs_if_zero(result[0].x) << ' ' << abs_if_zero(result[0].y) << '\n';
            continue;
        }

        cout << abs_if_zero(result[0].x) << ' ' << abs_if_zero(result[0].y) << ' '
                << abs_if_zero(result[1].x) << ' ' << abs_if_zero(result[1].y) << '\n';
    }

    cout << flush;
    return 0;
}

File: lab/segmentintersection.py
import sys

# Does not work

class Vec2D:
    EPS = 1e-9 

    x: float

    y: float

    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y
    def __sub__(self, other: "Vec2D") -> "Vec2D":
        return Vec2D(self.x - other.x, self.y - other.y)
    def __add__(self, other: "Vec2D") -> "Vec2D":
        return Vec2D(self.x + other.x, self.y + other.y)
    def __mul__(self, scalar: float) -> "Vec2D":
        return Vec2D(self.x * scalar, self.y * scalar)
    def __div__(self, scalar: float) -> "Vec2D":
        return Vec2D(self.x / scalar, self.y / scalar)
    def __eq__(self, other: "Vec2D") -> bool:
        return abs(self.x - other.x) < self.EPS and abs(self.y - other.y) < self.EPS

    def cross(self, other: "Vec2D") -> float:
        return self.x * other.y - self.y * other.x
    def dot(self, other: "Vec2D") -> float:
        return self.x * other.x + self.y * other.y
    def dist(self, other: "Vec2D") -> float:
        return ((self.x - other.x) ** 2 + (self.y - other.y) ** 2) ** 0.5

Point = Vec2D
Segment = tuple[Point, Point]

def intersect_col(s1: Segment, s2: Segment) -> list[Point]:
    u = s1[0] - s2[0]
    v = s1[0] - s1[1]

    # Lines are parallel
    if abs(v.cross(u)) > Vec2D.EPS:
        return []

    in_between = lambda a, b, x : a <= x <= b or b <= x <= a
    
    if not in_between(s1[0].x, s1[1].x, s2[0].x) and \
        not in_between(s1[0].x, s1[1].x, s2[1].x) and \
        not in_between(s2[0].x, s2[1].x, s1[0].x) and \
        not in_between(s2[0].x, s2[1].x, s1[1].x):
        return []

    res = [s1[0], s1[1], s2[0], s2[1]]
    res.sort(key=lambda p: p.x)

    res = res[1:]
    res = res[:-1]

    if res[0].dist(res[1]) < Vec2D.EPS:
        return [res[0]]

    return res

def intersect(s1: Segment, s2: Segment) -> list[Point]:
    p = s1[0]
    r = s1[1] - s1[0]

    q = s2[0]
    s = s2[1] - s2[0]

    rXs = r.cross(s)
    if abs(rXs) < Vec2D.EPS:
        return intersect_col(s1, s2)

    pq = q - p
    t = pq.cross(s) / rXs
    u = pq.cross(r) / rXs

    if not (0 <= t <= 1 and 0 <= u <= 1):
        return []

    res = p + r * t
    return [res]

if __name__ == "__main__":
    n = int(sys.stdin.readline().strip())
    for _ in range(n):
        x1, y1, x2, y2, x3, y3, x4, y4 = map(float, sys.stdin.readline().strip().split())

        s1 = (Point(x1, y1), Point(x2, y2))
        s2 = (Point(x3, y3), Point(x4, y4))

        res = intersect(s1, s2)
        if not res:
            print("none")
            continue

        if len(res) == 1:
            print(f"{res[0].x:.2f} {res[0].y:.2f}")
            continue

        print(f"{res[0].x:.2f} {res[0].y:.2f} {res[1].x:.2f} {res[1].y:.2f}")
